### PR TITLE
Patch/stereoupdates

### PR DESCRIPTION
**Pull request by johnmay**
Big pull request, sorry. Working with Bob (Jmol) on some stereochemistry work and needed some improvements/tweaks to the underlying CDK. Generally behaviour improvements/tweaks but now support extended cis/trans:

```
C/C=C=C=C/C
C/C=C=C=C\C
```

image

### CODE REVIEW

**Comment by egonw** [comment]
@rajarshi, @steinbeck, an extra pair of eyes is welcome. Please put in whatever time you have, if any.

File: base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java [hunk 1]
```diff
@@ -104,6 +104,9 @@
     /** ExtendedTetrahedral (e.g. 2,3-pentadiene) */
     public static final int AL   = 0x43_00;
 
+    /** ExtendedCisTrans (Cumulene) */
```
Comment (egonw) [inline base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java]: Very minor thing, cumulene is an example, right? -> "e.g. Cumulene", similar to line 104?

Comment (egonw) [reply base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java]: (ignore at will...)

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java [hunk 2]
```diff
@@ -113,36 +112,85 @@ protected StereoElementFactory(IAtomContainer container, int[][] graph, EdgeToBo
...
+        if (!atom.isInRing())
+            return false;
+        for (IBond bond : container.getConnectedBondsList(atom)) {
+            if (isInSmallRing(bond, max))
+                return true;
+        }
+        return false;
+    }
+
+    private IBond getOtherDb(IAtom a, IBond other) {
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: What does Db stand for here? Double bond, I guess?

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java [hunk 3]
```diff
@@ -365,6 +427,26 @@ private boolean isInSmallRing(IBond bond) {
...
+     *    C = C = C = C
+     *      ^   ^   ^
+     *      ^---^---^----- bonds
+     * </pre>
+     *
+     * @param bonds cumulated double bonds
+     * @param centers discovered stereocentres
+     * @return the extended cis/trans geometry if one could be created
+     */
+    abstract ExtendedCisTrans createExtendedCisTrans(List<IBond> bonds,
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: Maybe a silly question, but does the order of the IBond's in the list matter?

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java [hunk 4]
```diff
@@ -739,6 +824,63 @@ ExtendedTetrahedral createExtendedTetrahedral(int v, Stereocenters stereocenters
...
+            // only applies to odd-counts
+            if ((dbs.size() & 0x1) == 0)
+                return null;
+
+            IBond   focus    = dbs.get(dbs.size()/2);
+            IBond[] carriers = new IBond[2];
+            int     config   = 0;
+
+            IAtom begAtom = dbs.get(0).getOther(getShared(dbs.get(0), dbs.get(1)));
+            IAtom endAtom = dbs.get(dbs.size()-1).getOther(getShared(dbs.get(dbs.size()-1), dbs.get(dbs.size()-2)));
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: OK, so the passed List<IBond> when creating it must be sorted... I never remember which List implementations are guaranteed sorted and which not...

File: base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java (hunk 1 above)
Comment (johnmay) [reply base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java]: No... cumulene is the class of compounds.

https://en.wikipedia.org/wiki/Cumulene

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java (hunk 2 above)
Comment (johnmay) [reply base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: Yes

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java (hunk 3 above)
Comment (johnmay) [reply base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: Yes will update the doc there... but is a package private method.

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java (hunk 4 above)
Comment (johnmay) [reply base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: I don't know what you mean, sorted? I think you mean *ordered* which all lists are.

File: base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java (hunk 1 above)
Comment (egonw) [reply base/interfaces/src/main/java/org/openscience/cdk/interfaces/IStereoElement.java]: Reading that, there seems to be some controversy about the minimal number of double bonds, noting that heteroallenes are seen by some as (hetero)cumulenes too... :(

File: base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java (hunk 4 above)
Comment (egonw) [reply base/standard/src/main/java/org/openscience/cdk/stereo/StereoElementFactory.java]: What I mean is when the createExtendedCisTrans() is called, the past List<> must be ordered cumulatively... sounds reasonable, but might be good to say that in that method explicitly...

**Comment by johnmay** [comment]
Done

**Pull request merged**